tests/cleanup_and_retention_test.go

```diff
@@ -1,191 +1,191 @@
 			}
-			values := make([]row, 0, MaxPlaceholders)
-			for j := int64(0); j < total; j++ {
-				values = append(values, row{j, startMilli + j})
-
-				if len(values) == MaxPlaceholders || j == total-1 {
-					_, err := db.NamedExec(fmt.Sprintf(`INSERT INTO %s (id, event_time) VALUES (:id, :time)`, table), values)
-					require.NoError(t, err)
-					values = values[:0]
-				}
+		}
+	}
+
+	r := it.RedisServerT(t)
+	i := it.Icinga2NodeT(t, "master")
+	i.EnableIcingaDb(r)
+	i.Reload()
+	waitForDumpDoneSignal(t, r, 20*time.Second, 100*time.Millisecond)
+	config, err := yaml.Marshal(struct {
+		Retention retention `yaml:"history-retention"`
+	}{reten})
+	assert.NoError(t, err)
+	it.IcingaDbInstanceT(t, r, rdb, services.WithIcingaDbConfig(string(config)))
+
+	eventually.Assert(t, func(t require.TestingT) {
+		for category, stmt := range retentionStatements {
+			retentionDays, ok := reten.Options[category]
+			if !ok {
+				retentionDays = reten.Days
 			}
 
-			cleanup, err := db.CleanupOlderThan(
-				context.Background(),
-				icingadb.CleanupStmt{Table: table, Column: "event_time", PK: "id"},
-				uint64(scenario.limit),
-				start.Add(time.Millisecond*time.Duration(scenario.rowsToDelete)),
-			)
+			threshold := time.Now().AddDate(0, 0, -retentionDays)
+			thresholdMilli := threshold.UnixMilli()
+
+			var rowsLeft int
+			err := db.QueryRow(
+				db.Rebind(fmt.Sprintf(`SELECT COUNT(*) FROM %s WHERE %s < ?`, stmt.Table, stmt.Column)),
+				thresholdMilli,
+			).Scan(&rowsLeft)
 			assert.NoError(t, err)
 
-			var count uint64
-			err = db.QueryRow(fmt.Sprintf(`SELECT COUNT(*) FROM %s`, table)).Scan(&count)
+			var rowsSpared int
+			err = db.QueryRow(
+				db.Rebind(fmt.Sprintf(`SELECT COUNT(*) FROM %s WHERE %s >= ?`, stmt.Table, stmt.Column)),
+				thresholdMilli,
+			).Scan(&rowsSpared)
 			assert.NoError(t, err)
-			assert.Equal(t, int(scenario.rowsToSpare), int(count), "rows left in table")
-			assert.Equal(t, int(scenario.rowsToDelete), int(cleanup.Count), "deleted rows")
-			assert.Equal(t, int((scenario.rowsToDelete+scenario.limit-1)/scenario.limit), int(cleanup.Rounds), "number of rounds")
-		})
-	}
 
-	t.Run("History", func(t *testing.T) {
-		db := database(t)
-		retentionDays := 1
-		scenario := scenarios[len(scenarios)-1]
-		wg := sync.WaitGroup{}
-
-		for _, stmt := range history.RetentionStatements {
-			_, err := db.Exec(fmt.Sprintf(
-				`CREATE TABLE %[1]s (%[2]s int NOT NULL, %[3]s bigint NOT NULL, PRIMARY KEY (%[2]s))`,
-				stmt.Table, stmt.PK, stmt.Column))
-			require.NoError(t, err)
-
-			_, err = db.Exec(fmt.Sprintf(`CREATE INDEX idx_%[1]s ON %[1]s (%[2]s)`, stmt.Table, stmt.Column))
-			require.NoError(t, err)
-
-			total := scenario.rowsToDelete + scenario.rowsToSpare
-			type row struct {
-				Id   int64
-				Time int64
+			if retentionDays == 0 {
+				// No retention.
+				assert.Equal(t, rowsToDelete, rowsLeft, "rows left in retention period")
+				assert.Equal(t, rowsToSpare, rowsSpared, "rows spared")
```

Comment:
I'd just compare the sum of both in a single assert. `thresholdMilli` isn't really that much of a useful value to separate rows in this scenario. Also, if one of these asserts fails, I think it would be more confusing than just a single assert with a message like "all rows should still be present".